xgboost_ray/main.py

```diff
@@ -295,21 +340,44 @@ def train(self, rabit_args: List[str], params: Dict[str, Any],
         else:
             callbacks = []
         callbacks.append(self._save_checkpoint_callback())
+        callbacks.append(self._stop_training_callback())
         kwargs["callbacks"] = callbacks
 
-        with RabitContext(str(id(self)), rabit_args):
-            bst = xgb.train(
-                local_params,
-                local_dtrain,
-                *args,
-                evals=local_evals,
-                evals_result=evals_result,
-                **kwargs)
-            return {
-                "bst": bst,
-                "evals_result": evals_result,
-                "train_n": self._local_n
-            }
+        result_dict = {}
+
+        def _train():
+            try:
+                with RabitContext(str(id(self)), rabit_args):
+                    bst = xgb.train(
+                        local_params,
+                        local_dtrain,
+                        *args,
+                        evals=local_evals,
+                        evals_result=evals_result,
+                        **kwargs)
+                    result_dict.update({
+                        "bst": bst,
+                        "evals_result": evals_result,
+                        "train_n": self._local_n
+                    })
+            except XGBoostError:
+                # Silent fail, will be raised as RayXGBoostTrainingStopped
+                return
+
+        thread = threading.Thread(target=_train)
+        thread.daemon = True
+        thread.start()
+        while thread.is_alive():
+            thread.join(timeout=0)
+            if self._stop_event.is_set():
```

Comment:
In case of failure of actor A, this will stop training of the other actors (B, C, etc). If we didn't terminate the train() function (which we couldn't if we didn't use threads) the actors would be stuck and we couldn't start a new train() process on them

Reply:
Got it, makes sense! Do you think you can add this ^ as a comment in the code?